### PR TITLE
Enable Maven coordinates UI in flex wizard and generate pom.xml

### PR DESCRIPTION
**Pull request by chanseokoh**
First piece for #1716.

- Adds the Maven coordinates UI to the flex wizard. When enabled, it will just drop a single `pom.xml`.
- For the `pom.xml`, I based it on `pom.xml` from our HelloWorld archetype and made it a template file `pom.xml.flex.ftl`. 'mvn verify' on the command-line works at least, but I don't have great confidence as to whether its contents are exactly what we want.
- Coordinate validation works well in both the std wizard and the flex wizard. (As expected, validation is enabled only when the coordinate UI is enabled in the flex wizard.)

new app engine flexible project _019

### CODE REVIEW

File: plugins/com.google.cloud.tools.eclipse.appengine.newproject.test/src/com/google/cloud/tools/eclipse/appengine/newproject/CodeTemplatesTest.java [hunk 1]
```diff
@@ -85,6 +87,22 @@ public void testMaterializeAppEngineFlexFiles()
     validateAppYaml();
   }
 
+  @Test
+  public void testMaterializeAppEnigneFlexFiles_noPomXml() throws CoreException {
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.newproject.test/src/com/google/cloud/tools/eclipse/appengine/newproject/CodeTemplatesTest.java]: AppEngine (sp.)

File: plugins/com.google.cloud.tools.eclipse.appengine.newproject.test/src/com/google/cloud/tools/eclipse/appengine/newproject/CodeTemplatesTest.java [hunk 2]
```diff
@@ -85,6 +87,22 @@ public void testMaterializeAppEngineFlexFiles()
...
+    AppEngineProjectConfig config = new AppEngineProjectConfig();
+    CodeTemplates.materializeAppEngineFlexFiles(project, config, monitor);
+    assertFalse(project.getFile("pom.xml").exists());
+  }
+
+  @Test
+  public void testMaterializeAppEnigneFlexFiles_pomXmlIfEnablingMaven()
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.newproject.test/src/com/google/cloud/tools/eclipse/appengine/newproject/CodeTemplatesTest.java]: AppEngine

File: plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/AppEngineProjectConfig.java [hunk 3]
```diff
@@ -36,14 +36,19 @@
   private List<Library> appEngineLibraries = Collections.emptyList();
   private String serviceName;
 
+  private boolean asMavenProject;
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/AppEngineProjectConfig.java]: perhaps just useMaven or buildWithMaven

File: plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/AppEngineProjectWizard.java [hunk 4]
```diff
@@ -41,7 +41,7 @@
 public abstract class AppEngineProjectWizard extends Wizard implements INewWizard {
 
   private AppEngineWizardPage page = null;
-  private AppEngineProjectConfig config = new AppEngineProjectConfig();
+  protected AppEngineProjectConfig config = new AppEngineProjectConfig();
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/AppEngineProjectWizard.java]: final if possible. Maybe we need a getter here?

File: plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/AppEngineWizardPage.java [hunk 5]
```diff
@@ -160,11 +153,16 @@ private boolean validateLocalFields() {
     return true;
   }
 
-  private final class PageValidator implements ModifyListener {
+  public final class PageValidator implements ModifyListener, Listener {
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/AppEngineWizardPage.java]: move to an outer class

File: plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/CodeTemplates.java [hunk 6]
```diff
@@ -154,6 +154,17 @@ private static IFile materialize(IProject project, AppEngineProjectConfig config
 
     copyChildFile("favicon.ico", webapp, subMonitor.newChild(5)); //$NON-NLS-1$
 
+    if (config.getAsMavenProject()) {
+      Map<String, String> mavenCoordinates = new HashMap<>();
+      mavenCoordinates.put("projectGroupId", config.getMavenGroupId()); //$NON-NLS-1$
+      mavenCoordinates.put("projectArtifactId", config.getMavenArtifactId()); //$NON-NLS-1$
+      mavenCoordinates.put("projectVersion", config.getMavenVersion()); //$NON-NLS-1$
+      if (!isStandardProject) {
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/CodeTemplates.java]: This if shouldn't be necessary. For now a standard project should return false from getAsMavenProject.

File: plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/flex/AppEngineFlexWizardPage.java [hunk 7]
```diff
@@ -18,18 +18,41 @@
...
+  public boolean validatePage() {
+    if (!super.validatePage()) {
+      return false;
+    }
+    return mavenCoordinatesUi.setValidationMessage(this);
+  }
+
+  @Override
   public void sendAnalyticsPing(Shell shell) {
     // TODO: send analytics ping
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/flex/AppEngineFlexWizardPage.java]: not this PR, but we should get this done before launch if possible.

File: plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/maven/MavenCoordinatesUi.java [hunk 8]
```diff
@@ -99,10 +104,13 @@ public String getVersion() {
     return versionField.getText().trim();
   }
 
-  public void addModifyListener(ModifyListener listener) {
-    groupIdField.addModifyListener(listener);
-    artifactIdField.addModifyListener(listener);
-    versionField.addModifyListener(listener);
+  public void addChangeListener(Listener listener) {
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/maven/MavenCoordinatesUi.java]: I don't understand why this change is needed.

File: plugins/com.google.cloud.tools.eclipse.util/templates/appengine/pom.xml.flex.ftl [hunk 9]
```diff
@@ -0,0 +1,90 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.util/templates/appengine/pom.xml.flex.ftl]: remove copyright since we don't want to assert ownership of user's code

File: plugins/com.google.cloud.tools.eclipse.util/src/com/google/cloud/tools/eclipse/util/templates/appengine/AppEngineTemplateUtility.java [hunk 10]
```diff
@@ -41,6 +41,7 @@
   public static final String HELLO_APPENGINE_TEST_TEMPLATE = "HelloAppEngineTest.java.ftl";
   public static final String MOCK_HTTPSERVLETRESPONSE_TEMPLATE = "MockHttpServletResponse.java.ftl";
   public static final String APP_YAML_TEMPLATE = "app.yaml.ftl";
+  public static final String POM_XML_FLEX_TEMPLATE = "pom.xml.flex.ftl";
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.util/src/com/google/cloud/tools/eclipse/util/templates/appengine/AppEngineTemplateUtility.java]: we should change the standard pom.xml template name too

File: plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/maven/MavenCoordinatesUi.java (hunk 8 above)
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/maven/MavenCoordinatesUi.java]: I also need to add `PageValidator` (currently implementing `ModifierListener`) to the master control button for triggering validation when the value changes, but the button does not accept `ModifierListener`. Instead it accepts `SelectionListener`. So I decided to accept the generalized `Listener` version.

File: plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/AppEngineProjectWizard.java (hunk 4 above)
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/AppEngineProjectWizard.java]: `final` is possible. I feel OK without a getter: this is an `abstract` class, and only updating `config` are (one of) its subclasses.

File: plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/flex/AppEngineFlexWizardPage.java (hunk 7 above)
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/flex/AppEngineFlexWizardPage.java]: We have an issue for this: https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/1671

File: plugins/com.google.cloud.tools.eclipse.appengine.newproject.test/src/com/google/cloud/tools/eclipse/appengine/newproject/CodeTemplatesTest.java [hunk 11]
```diff
@@ -179,9 +208,9 @@ public void testCreateChildFile() throws CoreException, IOException {
         AppEngineTemplateUtility.HELLO_APPENGINE_TEMPLATE, parent, values, monitor);
     Assert.assertTrue(child.exists());
     Assert.assertEquals("HelloAppEngine.java", child.getName());
-    InputStream in = child.getContents(true);
-    try (BufferedReader reader = new BufferedReader(
-        new InputStreamReader(in, StandardCharsets.UTF_8.name()))) {
+    try (InputStream in = child.getContents(true);
+        BufferedReader reader = new BufferedReader(
+            new InputStreamReader(in, StandardCharsets.UTF_8.name()))) {
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.newproject.test/src/com/google/cloud/tools/eclipse/appengine/newproject/CodeTemplatesTest.java]: .name() is not needed here
